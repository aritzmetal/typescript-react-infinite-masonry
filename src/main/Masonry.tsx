import React, { CSSProperties, useEffect, useRef, useState } from "react"
import Bricks from "bricks.js"
import type { SizeDetail, Instance } from "bricks.js"
import InfiniteScroll, { Props as InfiniteScrollProps } from "react-infinite-scroll-component"

export interface Props extends InfiniteScrollProps {
    /** Classname to pass down styles */
    className?: string
    /** Update all options or just the new ones
     * @default false
     */
    pack?: boolean
    /** An attribute added to the grid items after they're positioned within the grid. */
    packed?: string
    /** A boolean, defaulting to true, indicating that the grid items should be positioned using the top and left CSS properties. */
    position?: boolean
    /** Media query and viewport breakpoints */
    sizes?: SizeDetail[]
    /** Custom */
    style?: CSSProperties
    /** Children of the masonry */
    children: React.ReactNode | JSX.Element
}

const defaultSizes: SizeDetail[] = [
    { columns: 1, gutter: 16 },
    { mq: "768px", columns: 2, gutter: 16 },
    { mq: "1024px", columns: 3, gutter: 16 },
    { mq: "1200px", columns: 4, gutter: 16 },
]

export const MasonryInfiniteScroller = ({
    className = "",
    pack = false,
    packed = className ? `${className}--packed` : `datum--packed`,
    position = true,
    sizes = defaultSizes,
    style = {},
    children,
    ...extraProps
}: Props) => {
    const ref = useRef<HTMLDivElement>(null)
    const [instance, setInstance] = useState<Instance>()

    useEffect(() => {
        if (ref.current) {
            setInstance(
                Bricks({
                    container: ref.current,
                    packed,
                    sizes,
                    position,
                })
            )
        }
    }, [ref.current])

    useEffect(() => {
        if (!!instance) {
            instance.pack()
            instance.resize(true)
        }
    }, [instance])

    useEffect(() => {
        if (instance) {
            if ((children as React.ReactNode[]).length > 0) {
                if (pack) {
                    instance.pack()
                } else {
                    instance.update()
                }
            }
        }
    }, [(children as React.ReactNode[]).length])

    return (
        <InfiniteScroll {...extraProps}>
            <div ref={ref} className={className} style={style}>
                {children}
            </div>
        </InfiniteScroll>
    )
}
