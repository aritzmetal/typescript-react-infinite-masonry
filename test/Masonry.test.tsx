import React from "react"
import { render, screen } from "@testing-library/react"
import "@testing-library/jest-dom"
import Masonry from "../src/index"

type Props = Parameters<typeof Masonry>[0]

describe("Masonry", () => {
    let wrapper: { baseElement: Element | Element[] }

    beforeEach(() => {
        Object.defineProperty(window, "matchMedia", {
            writable: true,
            value: jest.fn().mockImplementation((query) => ({
                matches: false,
                media: query,
                onchange: null,
                addListener: jest.fn(), // Deprecated
                removeListener: jest.fn(), // Deprecated
                addEventListener: jest.fn(),
                removeEventListener: jest.fn(),
                dispatchEvent: jest.fn(),
            })),
        })
    })

    const props: Props = {
        next: () => null,
        loader: null,
        hasMore: true,
        dataLength: 3,
        children: [1, 2, 3],
    }

    const mountWrapper = (extraProps?: Partial<Props>) => {
        if (extraProps) {
            wrapper = render(<Masonry {...props} {...extraProps} />)
        } else {
            wrapper = render(<Masonry {...props} />)
        }
    }
    test("Should match snapshot", () => {
        mountWrapper()

        expect(wrapper.baseElement).toMatchSnapshot("basic")
    })
})
