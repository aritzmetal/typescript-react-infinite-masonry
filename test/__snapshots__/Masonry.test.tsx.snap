// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Masonry Should match snapshot: basic 1`] = `
<body>
  <div>
    <div
      class="infinite-scroll-component__outerdiv"
    >
      <div
        class="infinite-scroll-component "
        style="height: auto; overflow: auto;"
      >
        <div
          class=""
          style="position: relative; height: -16px;"
        >
          1
          2
          3
        </div>
      </div>
    </div>
  </div>
</body>
`;
